File: 백준/Gold/2313. 보석 구매하기/보석 구매하기.py
import sys
input = sys.stdin.readline

n = int(input())
result_sum = 0
result_range = []

for _ in range(n):
    L = int(input())
    arr = list(map(int, input().split()))

    max_sum = arr[0]
    l, r = 0, 0
    ml, mr = 0, 0

    for i in range(1, L):
        if (arr[i] < arr[i-1] + arr[i]): # 누적합이 더 클 때
            arr[i] += arr[i-1]
            r = i
        else: # 손해인 경우, 리셋
            l = i
            r = i
        
        if (max_sum < arr[i]): # 최대치 갱신
            max_sum = arr[i]
            ml = l
            mr = r
        elif (max_sum == arr[i]): # 최대치가 같을 경우, 개수가 최소
            prev_range = mr - ml
            new_range = r - l

            if (prev_range > new_range):
                mr = r
                ml = l
    
    result_sum += max_sum
    result_range.append((ml + 1, mr + 1))

print(result_sum)
for result in result_range:
    print(*result)

File: 백준/Gold/2313. 보석 구매하기/보석 구매하기.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const n = Number(input[0]);
let tc = 0;
let result_sum = 0;
const result_range = [];

for (let i = 0; i < n; i++) {
    const L = Number(input[i + tc + 1]);
    let arr = input[i + tc + 2].split(" ").map(Number);
    tc++;

    let maxSum = arr[0];
    let maxL = 0;
    let maxR = 0;
    let l = 0;
    let r = 0;

    for (let j = 0; j < L; j++) {
        if (arr[j] < arr[j - 1] + arr[j]) {
            arr[j] += arr[j - 1];
            r = j;
        } else {
            l = j;
            r = j;
        }

        if (maxSum < arr[j]) {
            maxSum = arr[j];
            maxL = l;
            maxR = r;
        } else if (maxSum === arr[j]) {
            const prevCount = maxR - maxL;
            const newCount = r - l;

            if (prevCount > newCount) {
                maxL = l;
                maxR = r;
            }
        }
    }

    result_sum += maxSum;
    result_range.push([maxL + 1, maxR + 1]);
}

console.log(result_sum);
for (range of result_range) {
    console.log(range.join(" "));
}
